feat(lsp): include end_col and end_lnum in vim.lsp.buf.locations_to_items

tom-anders opened the pull request:
Noticed this by accident when after the Trouble v3 update, `:Trouble quickfix` would not display the correct range of items in the preview.

Might be worth a 0.10.0 backport? I assume many people upgraded to Trouble v3 and see this (currently only the first character of each item is highlighted in the preview, which looks a bit weird)

tom-anders committed:
feat(lsp): include end_col and end_lnum in vim.lsp.buf.locations_to_items

MariaSolOs commented on runtime/doc/news.txt:
```diff
@@ -113,6 +113,7 @@ LSP
 
 • Completion side effects (including snippet expansion, execution of commands
   and application of additional text edits) is now built-in.
+• |vim.lsp.util.locations_to_items()| correctly sets `end_col` field.
```
I wouldn't consider this a new feature (it almost feels like a bug fix), but I'm unsure.

mfussenegger replied:
I kinda disagree because `end_col` in qf items is relatively new.

clason replied:
Sorry, the "correctly" in the news threw me.

tom-anders commented:
Thanks for the quick feedback! I removed "correctly" from the news.txt entry, as discussed.

I also noticed that `end_lnum` was missing as well, so I added that as well.

tom-anders commented:
(I noticed that the calculation for `end_col` would possibly yield false results if `lnum!=end_lnum`, the last force-push fixes that and adapts the test accordingly

rockyzhang24 commented:
Great. And I think the doc for `vim.lsp.util.location_to_items()` should be updated as well.

tom-anders commented:
> Great. And I think the doc for `vim.lsp.util.locations_to_items()` (its `Return`) should be updated as well.

Done!

One more thing: Right now we set `end_col` and `end_lnum` even if they are actually the same as `col` and `lnum` - This doesn't seem to cause any problems, but the docs of `setqflist` at least seem to indicate that `end_col`/`end_lnum` should only be set if they're different from `col`/`lnum`. Is it worth changing (and complicating) the code for this?

rockyzhang24 commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -1760,7 +1762,7 @@ function M.locations_to_items(locations, offset_encoding)
     local uri = d.uri or d.targetUri
     local range = d.range or d.targetSelectionRange
-    table.insert(grouped[uri], { start = range.start, location = d })
+    table.insert(grouped[uri], { start = range.start, ['end'] = range['end'], location = d })
```
Should the annotation for `grouped` above be adjusted accordingly?
```diff
- ---@type table<string, {start: lsp.Position, location: lsp.Location|lsp.LocationLink}[]>
+ ---@type table<string, {start: lsp.Position, end: lsp.Position, location: lsp.Location|lsp.LocationLink}[]>
```

ofseed commented:
Found this is duplicated with https://github.com/neovim/neovim/pull/28724. Feel free to close it if you decided to merge this one.

tom-anders commented:
whoops, I force pushed to add the type annotation suggested by @rockyzhang24 , but accidentally reverted another change (adding `end_col` and `end_lnum` to the docs) - The second force push fixes this. So, looks like this is ready to be merged?
